medusa/backup_cluster.py

```diff
@@ -0,0 +1,165 @@
+
+        logging.debug('Done emitting metrics.')
+        logging.info('Backup of the cluster done.')
+
+    except Exception as e:
+        tags = ['medusa-cluster-backup', 'cluster-backup-error', backup_name]
+        monitoring.send(tags, 1)
+
+        logging.error('This error happened during the cluster backup: {}'.format(str(e)))
+        traceback.print_exc()
+
+        pssh_run_success_cleanup = backup.orchestration_custom \
+            .pssh_run(backup.hosts,
+                      backup.cassandra.delete_snapshot_command(backup.snapshot_tag),
+                      hosts_variables={})
+        if not pssh_run_success_cleanup:
+            err_msg_cleanup = 'Some nodes failed to clear the snapshot. Cleaning snapshots manually is recommended'
+            logging.error(err_msg_cleanup)
+        sys.exit(1)
+
+
+class BackupJob(object):
+    def __init__(self, config, backup_name, stagger, mode, seed_target, pssh_pool_size):
+        self.id = uuid.uuid4()
+        self.orchestration_snapshots = Orchestration(config, 1000)
+        self.orchestration_custom = Orchestration(config, pssh_pool_size)
+        self.config = config
+        self.backup_name = backup_name
+        self.stagger = stagger
+        self.mode = mode
+        self.seed_target = seed_target
+        self.pssh_pool_size = pssh_pool_size
+        self.work_dir = self.temp_dir / 'medusa-job-{id}'.format(id=self.id)
+        self.hosts = {}
+        self.cassandra = Cassandra(config.cassandra)
+        self.snapshot_tag = '{}{}'.format(self.cassandra.Snapshot.prefix, self.backup_name)
+        fqdn_resolver = medusa.config.evaluate_boolean(self.config.cassandra.resolve_ip_addresses)
+        self.fqdn_resolver = HostnameResolver(fqdn_resolver)
+
+    def execute(self):
+        # Two step: Take snapshot everywhere, then upload the backups to the external storage
+
+        # Geeting
+        session_provider = CqlSessionProvider([self.seed_target],
+                                              self.config.cassandra)
+        with session_provider.new_session() as session:
+            tokenmap = session.tokenmap()
+            self.hosts = tokenmap.values()
+
+        # First let's take a snapshot on all nodes at once
+        # Here we will use pssh_pool_size = number of nodes
+        logging.info('Creating snapshots on all nodes')
+        self._create_snapshots(self)
+
+        # Second
+        logging.info('Uploading snapshots from nodes to external storage')
+        self._upload_backup(self)
+
+    def _create_snapshots(self):
+        # Run snapshot in parallel on all nodes,
+        pssh_run_success = self.orchestration_snapshots \
+            .pssh_run(self.hosts,
+                      self.cassandra.create_snapshot_command(self.backup_name),
+                      hosts_variables={})
+        if not pssh_run_success:
+            # we could implement a retry.
+            err_msg = 'Some nodes failed to create the snapshot. Cleaning snapshots and exiting'
```

Comment:
I don't think that message is accurate as we're not doing cleanups. Are we?

Reply:
That's because I moved the cleanup of the snapshot on the execute exception handle at the last moment. This comment needs to be split and the part about cleaning snapshot needs to go where we actually do it (same place where we return a metric saying backups failed). Good catch.